File: sdk-go/verify.go
package ztxbas

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
	"sync"
	"time"
)

// Claims are the fields ztxbas includes in every issued JWT.
type Claims struct {
	Issuer      string `json:"iss,omitempty"`
	Subject     string `json:"sub,omitempty"`
	Audience    string `json:"aud,omitempty"`
	IssuedAt    int64  `json:"iat,omitempty"`
	ExpiresAt   int64  `json:"exp,omitempty"`
	Email       string `json:"email,omitempty"`
	Origin      string `json:"origin,omitempty"`
	ChallengeID string `json:"challenge_id,omitempty"`
}

// jwkSet is the minimal JWKS shape we consume. Extra fields are ignored.
type jwkSet struct {
	Keys []jwk `json:"keys"`
}

type jwk struct {
	Kty string `json:"kty"`
	Crv string `json:"crv"`
	Alg string `json:"alg"`
	Use string `json:"use"`
	Kid string `json:"kid"`
	X   string `json:"x"`
	Y   string `json:"y"`
}

// jwksCacheTTL is how long a fetched JWKS is trusted before a fresh
// fetch. Short enough to propagate a rotation in a few minutes without
// operator intervention, long enough that spike traffic doesn't hammer
// the JWKS endpoint.
const jwksCacheTTL = 5 * time.Minute

// clockSkew is the leeway allowed on exp/iat/nbf when verifying tokens.
// Same value the server uses for HMAC timestamps, kept in sync so a JWT
// minted at t=now is not rejected by a caller whose clock is a few
// seconds behind.
const clockSkew = 30 * time.Second

// Verifier verifies ztxbas-issued JWTs against a cached JWKS.
//
// Constructed automatically by Client.VerifyJWT on first use, or
// explicitly via NewVerifier for callers who only want the verify side
// of the SDK (e.g. an RP backend that receives JWTs from a mobile
// frontend and doesn't itself call the ztxbas API).
type Verifier struct {
	jwksURL string
	hc      *http.Client

	mu          sync.Mutex
	keys        map[string]*ecdsa.PublicKey // kid → key
	fetchedAt   time.Time
	expectedIss string // optional; if set, JWTs must carry iss==this
}

// VerifierOption configures a Verifier.
type VerifierOption func(*Verifier)

// WithExpectedIssuer requires that verified JWTs carry the given `iss`.
// Recommended for RPs that talk to a single ztxbas deployment.
func WithExpectedIssuer(iss string) VerifierOption {
	return func(v *Verifier) { v.expectedIss = iss }
}

// WithVerifierHTTPClient overrides the HTTP client used for JWKS fetches.
func WithVerifierHTTPClient(hc *http.Client) VerifierOption {
	return func(v *Verifier) { v.hc = hc }
}

// NewVerifier builds a standalone Verifier from a JWKS URL.
// The typical URL is `<baseURL>/.well-known/jwks.json`.
func NewVerifier(jwksURL string, opts ...VerifierOption) (*Verifier, error) {
	if jwksURL == "" {
		return nil, fmt.Errorf("ztxbas: jwksURL required")
	}
	v := &Verifier{
		jwksURL: jwksURL,
		hc:      &http.Client{Timeout: DefaultTimeout},
	}
	for _, o := range opts {
		o(v)
	}
	return v, nil
}

// VerifyJWT parses and verifies token, returning the claims on success.
// Verification includes signature (ES256/P-256), exp, iat, and — if a
// Verifier was configured with WithExpectedIssuer — iss.
//
// The convenience method on Client lazily builds a Verifier pointing at
// the same base URL, so a call to Client.VerifyJWT is enough for most
// integrations.
func (c *Client) VerifyJWT(ctx context.Context, token string) (*Claims, error) {
	if c.verifier == nil {
		// Race is benign — two goroutines both making a verifier just
		// duplicate the initial JWKS fetch; caching is per-instance so
		// the second will discard its work on next Verify.
		v, err := NewVerifier(c.baseURL.String()+"/.well-known/jwks.json",
			WithVerifierHTTPClient(c.hc))
		if err != nil {
			return nil, err
		}
		c.verifier = v
	}
	return c.verifier.Verify(ctx, token)
}

// Verify parses and verifies token, refreshing the JWKS if needed.
func (v *Verifier) Verify(ctx context.Context, token string) (*Claims, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("ztxbas: malformed JWT: want 3 parts, got %d", len(parts))
	}
	headerBytes, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("ztxbas: decode JWT header: %w", err)
	}
	var hdr struct {
		Alg string `json:"alg"`
		Typ string `json:"typ"`
		Kid string `json:"kid"`
	}
	if err := json.Unmarshal(headerBytes, &hdr); err != nil {
		return nil, fmt.Errorf("ztxbas: parse JWT header: %w", err)
	}
	// Reject anything other than the algorithm we know. This is the
	// classic "alg=none" / algorithm-substitution defence.
	if hdr.Alg != "ES256" {
		return nil, fmt.Errorf("ztxbas: unsupported JWT alg %q (want ES256)", hdr.Alg)
	}
	if hdr.Kid == "" {
		return nil, errors.New("ztxbas: JWT header missing kid")
	}

	key, err := v.keyForKID(ctx, hdr.Kid)
	if err != nil {
		return nil, err
	}

	sigBytes, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return nil, fmt.Errorf("ztxbas: decode JWT signature: %w", err)
	}
	// ES256 signature is r||s, each 32 bytes big-endian.
	if len(sigBytes) != 64 {
		return nil, fmt.Errorf("ztxbas: ES256 signature must be 64 bytes, got %d", len(sigBytes))
	}
	r := new(big.Int).SetBytes(sigBytes[:32])
	s := new(big.Int).SetBytes(sigBytes[32:])

	signingInput := parts[0] + "." + parts[1]
	digest := sha256.Sum256([]byte(signingInput))
	if !ecdsa.Verify(key, digest[:], r, s) {
		return nil, errors.New("ztxbas: JWT signature invalid")
	}

	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("ztxbas: decode JWT payload: %w", err)
	}
	var claims Claims
	if err := json.Unmarshal(payloadBytes, &claims); err != nil {
		return nil, fmt.Errorf("ztxbas: parse JWT payload: %w", err)
	}

	now := time.Now().Unix()
	if claims.ExpiresAt != 0 && now > claims.ExpiresAt+int64(clockSkew.Seconds()) {
		return nil, fmt.Errorf("ztxbas: JWT expired at %d (now %d)", claims.ExpiresAt, now)
	}
	if claims.IssuedAt != 0 && claims.IssuedAt > now+int64(clockSkew.Seconds()) {
		return nil, fmt.Errorf("ztxbas: JWT iat in the future (%d, now %d)", claims.IssuedAt, now)
	}
	if v.expectedIss != "" && claims.Issuer != v.expectedIss {
		return nil, fmt.Errorf("ztxbas: JWT iss %q does not match expected %q", claims.Issuer, v.expectedIss)
	}
	return &claims, nil
}

// keyForKID returns the cached key for kid, refreshing the JWKS if the
// cache is stale or the kid is unknown.
func (v *Verifier) keyForKID(ctx context.Context, kid string) (*ecdsa.PublicKey, error) {
	v.mu.Lock()
	if key, ok := v.keys[kid]; ok && time.Since(v.fetchedAt) < jwksCacheTTL {
		v.mu.Unlock()
		return key, nil
	}
	v.mu.Unlock()

	// Fetch outside the lock so a slow JWKS endpoint doesn't stall
	// concurrent verifies waiting on the cache.
	fresh, err := v.fetchJWKS(ctx)
	if err != nil {
		return nil, err
	}
	v.mu.Lock()
	v.keys = fresh
	v.fetchedAt = time.Now()
	key, ok := v.keys[kid]
	v.mu.Unlock()
	if !ok {
		return nil, fmt.Errorf("ztxbas: JWT kid %q not present in JWKS", kid)
	}
	return key, nil
}

// fetchJWKS pulls the JWKS document and decodes it into a kid→key map.
// Bounded by a 64 KiB read cap — legitimate JWKS docs are ≤1 KiB.
func (v *Verifier) fetchJWKS(ctx context.Context) (map[string]*ecdsa.PublicKey, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, v.jwksURL, nil)
	if err != nil {
		return nil, fmt.Errorf("ztxbas: build JWKS request: %w", err)
	}
	req.Header.Set("Accept", "application/json")
	resp, err := v.hc.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ztxbas: fetch JWKS: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("ztxbas: JWKS fetch returned %d", resp.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, 64<<10))
	if err != nil {
		return nil, fmt.Errorf("ztxbas: read JWKS: %w", err)
	}
	var set jwkSet
	if err := json.Unmarshal(body, &set); err != nil {
		return nil, fmt.Errorf("ztxbas: parse JWKS: %w", err)
	}
	out := make(map[string]*ecdsa.PublicKey, len(set.Keys))
	for _, k := range set.Keys {
		if k.Kty != "EC" || k.Crv != "P-256" {
			continue // skip anything we can't verify
		}
		pk, err := jwkToECDSA(k)
		if err != nil {
			return nil, fmt.Errorf("ztxbas: decode key %q: %w", k.Kid, err)
		}
		out[k.Kid] = pk
	}
	if len(out) == 0 {
		return nil, errors.New("ztxbas: JWKS has no usable P-256 keys")
	}
	return out, nil
}

// jwkToECDSA rebuilds an ecdsa.PublicKey from a JWK's x/y coordinates.
// Validates that the point is on the P-256 curve — an off-curve key
// would let an attacker forge signatures via invalid-curve attacks.
func jwkToECDSA(k jwk) (*ecdsa.PublicKey, error) {
	x, err := base64.RawURLEncoding.DecodeString(k.X)
	if err != nil {
		return nil, fmt.Errorf("x: %w", err)
	}
	y, err := base64.RawURLEncoding.DecodeString(k.Y)
	if err != nil {
		return nil, fmt.Errorf("y: %w", err)
	}
	curve := elliptic.P256()
	pk := &ecdsa.PublicKey{
		Curve: curve,
		X:     new(big.Int).SetBytes(x),
		Y:     new(big.Int).SetBytes(y),
	}
	if !curve.IsOnCurve(pk.X, pk.Y) {
		return nil, errors.New("public key not on P-256 curve")
	}
	return pk, nil
}

File: sdk-python/ztxbas/verify.py
"""JWT verification against the ztxbas JWKS.

Uses ``cryptography`` for ES256 primitives. Kept dependency-light: no
PyJWT, no third-party JWKS lib.
"""

from __future__ import annotations

import base64
import json
import threading
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Callable, Dict, Optional

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import encode_dss_signature

from .errors import JwtVerifyError

#: How long a cached JWKS document is trusted before re-fetching.
_JWKS_CACHE_TTL = 5 * 60  # seconds

#: Leeway on ``iat``/``exp`` — matches server-side HMAC timestamp skew.
_CLOCK_SKEW = 30  # seconds


@dataclass
class Claims:
    """Fields ztxbas includes in every issued JWT.

    All fields are optional at the type level because JWTs from other
    issuers, or malformed ones, might omit them. In practice a
    successful ``verify()`` call always populates ``iss``, ``sub``,
    ``exp``, ``email``, ``origin``, and ``challenge_id``.
    """

    iss: Optional[str] = None
    sub: Optional[str] = None
    aud: Optional[str] = None
    iat: Optional[int] = None
    exp: Optional[int] = None
    email: Optional[str] = None
    origin: Optional[str] = None
    challenge_id: Optional[str] = None

    @classmethod
    def from_dict(cls, d: Dict[str, object]) -> "Claims":
        return cls(
            iss=d.get("iss"),  # type: ignore[arg-type]
            sub=d.get("sub"),  # type: ignore[arg-type]
            aud=d.get("aud"),  # type: ignore[arg-type]
            iat=d.get("iat"),  # type: ignore[arg-type]
            exp=d.get("exp"),  # type: ignore[arg-type]
            email=d.get("email"),  # type: ignore[arg-type]
            origin=d.get("origin"),  # type: ignore[arg-type]
            challenge_id=d.get("challenge_id"),  # type: ignore[arg-type]
        )


def _b64url_decode(s: str) -> bytes:
    """Decode base64url, tolerating missing padding (per RFC 7515)."""
    pad = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + pad)


class Verifier:
    """Verifies ztxbas-issued JWTs against a cached JWKS.

    Thread-safe. JWKS fetches are coalesced under a lock so a burst of
    verifies against a cold cache produces at most one HTTP call.
    """

    def __init__(
        self,
        jwks_url: str,
        expected_issuer: Optional[str] = None,
        http_get: Optional[Callable[[str], bytes]] = None,
        timeout: float = 10.0,
    ) -> None:
        """Build a verifier.

        Args:
            jwks_url: Fully-qualified URL, typically
                ``https://<host>/.well-known/jwks.json``.
            expected_issuer: If set, verified JWTs must carry ``iss``
                equal to this value. Recommended for RPs talking to a
                single ztxbas deployment.
            http_get: Callable ``(url) -> bytes`` used to fetch the
                JWKS. Defaults to ``urllib.request``; tests can inject
                a fake.
            timeout: Timeout for the default HTTP fetcher, in seconds.
        """
        if not jwks_url:
            raise ValueError("ztxbas: jwks_url required")
        self._jwks_url = jwks_url
        self._expected_issuer = expected_issuer
        self._timeout = timeout
        self._http_get = http_get or self._default_get
        self._lock = threading.Lock()
        self._keys: Dict[str, ec.EllipticCurvePublicKey] = {}
        self._fetched_at: float = 0.0

    # ---- HTTP -----------------------------------------------------------

    def _default_get(self, url: str) -> bytes:
        req = urllib.request.Request(url, headers={"Accept": "application/json"})
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                if resp.status != 200:
                    raise JwtVerifyError(f"JWKS fetch returned {resp.status}")
                # Cap the read — legitimate JWKS docs are ~1 KiB.
                return resp.read(65536)
        except urllib.error.URLError as e:
            raise JwtVerifyError(f"fetch JWKS: {e}") from e

    # ---- public API -----------------------------------------------------

    def verify(self, token: str) -> Claims:
        """Verify ``token``. Return the claims on success, raise on failure."""
        parts = token.split(".")
        if len(parts) != 3:
            raise JwtVerifyError(
                f"malformed JWT: expected 3 parts, got {len(parts)}"
            )
        enc_header, enc_payload, enc_sig = parts
        try:
            header = json.loads(_b64url_decode(enc_header))
        except (ValueError, json.JSONDecodeError) as e:
            raise JwtVerifyError(f"decode JWT header: {e}") from e

        # Reject anything other than ES256 up front — the classic
        # alg=none / algorithm-substitution defence.
        if header.get("alg") != "ES256":
            raise JwtVerifyError(
                f"unsupported JWT alg {header.get('alg')!r} (want ES256)"
            )
        kid = header.get("kid")
        if not kid:
            raise JwtVerifyError("JWT header missing kid")

        key = self._key_for_kid(kid)

        try:
            raw_sig = _b64url_decode(enc_sig)
        except (ValueError, TypeError) as e:
            raise JwtVerifyError(f"decode JWT signature: {e}") from e
        if len(raw_sig) != 64:
            raise JwtVerifyError(
                f"ES256 signature must be 64 bytes, got {len(raw_sig)}"
            )
        # cryptography expects a DER-encoded (r, s) pair; convert from
        # the raw r||s that JWS puts on the wire (RFC 7515 §3.4).
        r = int.from_bytes(raw_sig[:32], "big")
        s = int.from_bytes(raw_sig[32:], "big")
        der_sig = encode_dss_signature(r, s)

        signing_input = f"{enc_header}.{enc_payload}".encode("ascii")
        try:
            key.verify(der_sig, signing_input, ec.ECDSA(hashes.SHA256()))
        except InvalidSignature as e:
            raise JwtVerifyError("JWT signature invalid") from e

        try:
            payload = json.loads(_b64url_decode(enc_payload))
        except (ValueError, json.JSONDecodeError) as e:
            raise JwtVerifyError(f"decode JWT payload: {e}") from e

        now = int(time.time())
        exp = payload.get("exp")
        if exp is not None and now > exp + _CLOCK_SKEW:
            raise JwtVerifyError(f"JWT expired at {exp} (now {now})")
        iat = payload.get("iat")
        if iat is not None and iat > now + _CLOCK_SKEW:
            raise JwtVerifyError(f"JWT iat in the future ({iat}, now {now})")
        if self._expected_issuer and payload.get("iss") != self._expected_issuer:
            raise JwtVerifyError(
                f"JWT iss {payload.get('iss')!r} does not match "
                f"expected {self._expected_issuer!r}"
            )
        return Claims.from_dict(payload)

    # ---- internal -------------------------------------------------------

    def _key_for_kid(self, kid: str) -> ec.EllipticCurvePublicKey:
        with self._lock:
            key = self._keys.get(kid)
            if key is not None and time.time() - self._fetched_at < _JWKS_CACHE_TTL:
                return key
            self._refresh_locked()
            key = self._keys.get(kid)
            if key is None:
                raise JwtVerifyError(f"JWT kid {kid!r} not present in JWKS")
            return key

    def _refresh_locked(self) -> None:
        raw = self._http_get(self._jwks_url)
        try:
            doc = json.loads(raw)
        except (ValueError, json.JSONDecodeError) as e:
            raise JwtVerifyError(f"parse JWKS: {e}") from e
        keys = doc.get("keys")
        if not isinstance(keys, list):
            raise JwtVerifyError("JWKS missing keys array")
        out: Dict[str, ec.EllipticCurvePublicKey] = {}
        for k in keys:
            if k.get("kty") != "EC" or k.get("crv") != "P-256":
                continue
            try:
                x = int.from_bytes(_b64url_decode(k["x"]), "big")
                y = int.from_bytes(_b64url_decode(k["y"]), "big")
                # from_encoded_point raises for off-curve points, which
                # is exactly the invalid-curve defence we want.
                pub_numbers = ec.EllipticCurvePublicNumbers(x, y, ec.SECP256R1())
                out[k["kid"]] = pub_numbers.public_key()
            except (KeyError, ValueError) as e:
                raise JwtVerifyError(
                    f"decode key {k.get('kid')!r}: {e}"
                ) from e
        if not out:
            raise JwtVerifyError("JWKS has no usable P-256 keys")
        self._keys = out
        self._fetched_at = time.time()
